clippy_lints/src/casts/mod.rs

```diff
@@ -555,6 +556,47 @@ declare_clippy_lint! {
     "detects `as _` conversion"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for the usage of `as *const _` or `as *mut _` where the pointed to type is inferred.
+    ///
+    /// ### Why is this bad?
+    /// When converting to a pointer, it can be dangerous to not specify the type. Type inference can
+    /// be affected by many things, types may not always be obvious, and when working with pointers, while
+    /// the pointed to type doesn't technically matter until an access, you should always know what types
+    /// you intend to work with. When using multiple chained `as` casts, it can be especially dangerous,
+    /// because `*const T as *const U` **always compiles** (for Sized types T and U).
+    ///
+    /// ### Example
+    /// ```rust
+    /// struct UwU;
+    /// impl UwU {
+    ///     // intent: turn a `&UwU` into a `*const u8` that points to the same data
+    ///     fn as_ptr(&self) -> *const u8 {
+    ///         // ⚠️ `&self` is a `&&UwU`, so this turns a double pointer into a single pointer
+    ///         // ⚠️ This pointer is a dangling pointer to a local
+    ///         &self as *const _ as *const u8
+    ///     }
+    /// }
```

Comment:
How about using something like this as our example? Has a non buggy case to contrast to and it's pretty plain since there's no ZSTs/self params to think about

```rust
fn good_cast(x: &u64) -> *const u8 {
    x as *const _ as *const u8
}

fn bad_cast(x: &u64) -> *const u8 {
    &x as *const _ as *const u8
}
```